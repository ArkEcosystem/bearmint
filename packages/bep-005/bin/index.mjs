#!/usr/bin/env node
import { join } from 'node:path'

import { globby } from 'globby'
import meow from 'meow'

process.env.NODE_ENV = 'test'

const { flags } = meow({
	flags: {
		config: {
			type: 'string',
		},
	},
	importMeta: import.meta,
})

const { files } = (await import(join(process.cwd(), flags.config))).default
const tests = (await globby(files)).map((path) => join(process.cwd(), path))

for (const test of tests) {
	await import(test)
}
