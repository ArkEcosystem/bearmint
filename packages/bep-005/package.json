{
	"name": "@bearmint/bep-005",
	"version": "0.0.0",
	"private": true,
	"description": "This library is the official reference implementation for BEP-005 (https://bep.dev/005)",
	"keywords": [
		"bearmint",
		"tendermint",
		"blockchain",
		"framework",
		"modular"
	],
	"homepage": "https://github.com/bearmint/bearmint",
	"repository": {
		"type": "git",
		"url": "git+https://github.com/bearmint/bearmint.git"
	},
	"funding": "https://funding.bearmint.com/",
	"license": "MIT",
	"author": "Bearmint Mining Company <hello@bearmint.com> (https://bearmint.com/)",
	"type": "module",
	"exports": "./distribution/index.js",
	"types": "./distribution/index.d.ts",
	"bin": {
		"roadrunner": "./bin/index.mjs"
	},
	"files": [
		"bin",
		"distribution"
	],
	"scripts": {
		"audit": "pnpm audit --fix --prod",
		"build": "pnpm run clean && tsc",
		"build:watch": "pnpm run clean && tsc -w",
		"clean": "rm -rf distribution",
		"depcheck": "depcheck --ignores=@bearmint/bep-004,@bearmint/bep-005,@bearmint/bep-031,@bearmint/bep-039,globby,meow",
		"deps:major": "ncu -u --target=latest",
		"deps:minor": "ncu -u --target=minor",
		"deps:patch": "ncu -u --target=patch"
	},
	"dependencies": {
		"@bearmint/bep-013": "workspace:^",
		"@bearmint/bep-018": "workspace:^",
		"@bearmint/bep-116": "workspace:^",
		"@sinonjs/fake-timers": "9.1.2",
		"ansi-styles": "6.2.1",
		"concordance": "5.0.4",
		"globby": "13.1.2",
		"lodash.get": "4.4.2",
		"lodash.has": "4.5.2",
		"meow": "^11.0.0",
		"nock": "13.2.9",
		"pretty-ms": "8.0.0",
		"sinon": "14.0.1",
		"strip-ansi": "7.0.1",
		"vitest": "^0.24.5"
	},
	"devDependencies": {
		"@types/node": "18.11.9",
		"@types/sinon": "10.0.13",
		"@types/sinonjs__fake-timers": "8.1.2"
	},
	"engines": {
		"node": ">=18",
		"pnpm": ">=7.9.0"
	},
	"publishConfig": {
		"access": "restricted"
	}
}
