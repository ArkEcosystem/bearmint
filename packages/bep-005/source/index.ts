export * from './expect.js'
export * from './fake.js'
export * from './nock.js'
export * from './spy.js'
export * from './stub.js'
