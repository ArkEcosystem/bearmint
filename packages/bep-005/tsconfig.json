{
	"compilerOptions": {
		"typeRoots": ["node_modules/@types"],
		"outDir": "distribution"
	},
	"exclude": ["node_modules", "test/**/**.ts"],
	"extends": "../../config/tsconfig.json",
	"include": ["source/**/**.ts"]
}
