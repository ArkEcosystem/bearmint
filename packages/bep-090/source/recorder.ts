import type { EventRecorder, EventRecorderAttribute, EventRecorderFactory } from '@bearmint/bep-013'
import { abci } from '@bearmint/bep-018'

export function makeEventRecorder(): EventRecorder {
	let errors: abci.EventAttribute[] = []
	let events: abci.EventAttribute[] = []
	let type: string

	function canonicalizeAttributes(attribute: EventRecorderAttribute) {
		return new abci.EventAttribute({
			index: attribute.index ?? false,
			key: attribute.key,
			value: attribute.value,
		})
	}

	function captureError(attribute: EventRecorderAttribute) {
		errors.push(canonicalizeAttributes(attribute))
	}

	function captureEvent(attribute: EventRecorderAttribute) {
		events.push(canonicalizeAttributes(attribute))
	}

	return {
		actingAs(value: string) {
			errors = []
			events = []
			type = value
		},
		all() {
			return new abci.Event({ attributes: [...errors, ...events], type })
		},
		captureError,
		captureErrors(attributes: EventRecorderAttribute[]) {
			for (const attribute of attributes) {
				captureError(attribute)
			}

			return attributes.length > 0
		},
		captureEvent,
		captureEvents(attributes: EventRecorderAttribute[]) {
			for (const attribute of attributes) {
				captureEvent(attribute)
			}

			return attributes.length > 0
		},
		errors() {
			return new abci.Event({
				attributes: errors,
				type,
			})
		},
		events() {
			return new abci.Event({
				attributes: events,
				type,
			})
		},
		hasErrors() {
			return errors.length > 0
		},
		hasEvents() {
			return events.length > 0
		},
	}
}

export function makeEventRecorderFactory(): EventRecorderFactory {
	return {
		make() {
			return makeEventRecorder()
		},
	}
}
