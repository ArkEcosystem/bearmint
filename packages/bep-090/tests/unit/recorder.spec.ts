import { expect, describe, it, beforeEach } from 'vitest'
import { extendExpect } from '@bearmint/bep-005'
extendExpect(expect)
import type { EventRecorder } from '@bearmint/bep-013'

import { makeEventRecorder } from '../../source/recorder.js'

describe<{
	subject: EventRecorder
}>('Event Recorder', () => {
	beforeEach((context) => {
		context.subject = makeEventRecorder()
	})

	it('should capture errors', async (context) => {
		context.subject.actingAs('type')

		expect(context.subject.hasErrors()).toBeFalse()

		context.subject.captureError({ key: 'key1', value: 'key1' })
		context.subject.captureErrors([{ key: 'key2', value: 'key2' }])
		context.subject.captureErrors([{ key: 'key3', value: 'key3' }])

		expect(context.subject.hasErrors()).toBeTrue()

		expect(context.subject.errors()).toMatchSnapshot()
		expect(context.subject.all()).toMatchSnapshot()
	})

	it('should capture events', async (context) => {
		context.subject.actingAs('type')

		expect(context.subject.hasEvents()).toBeFalse()

		context.subject.captureEvent({ key: 'key1', value: 'key1' })
		context.subject.captureEvents([{ key: 'key2', value: 'key2' }])
		context.subject.captureEvents([{ key: 'key3', value: 'key3' }])

		expect(context.subject.hasEvents()).toBeTrue()

		expect(context.subject.events()).toMatchSnapshot()
		expect(context.subject.all()).toMatchSnapshot()
	})
})
