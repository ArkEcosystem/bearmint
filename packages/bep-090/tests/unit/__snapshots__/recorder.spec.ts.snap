// Vitest Snapshot v1

exports[`Event Recorder > should capture errors 1`] = `
Event {
  "attributes": [
    EventAttribute {
      "index": false,
      "key": "key1",
      "value": "key1",
    },
    EventAttribute {
      "index": false,
      "key": "key2",
      "value": "key2",
    },
    EventAttribute {
      "index": false,
      "key": "key3",
      "value": "key3",
    },
  ],
  "type": "type",
}
`;

exports[`Event Recorder > should capture errors 2`] = `
Event {
  "attributes": [
    EventAttribute {
      "index": false,
      "key": "key1",
      "value": "key1",
    },
    EventAttribute {
      "index": false,
      "key": "key2",
      "value": "key2",
    },
    EventAttribute {
      "index": false,
      "key": "key3",
      "value": "key3",
    },
  ],
  "type": "type",
}
`;

exports[`Event Recorder > should capture events 1`] = `
Event {
  "attributes": [
    EventAttribute {
      "index": false,
      "key": "key1",
      "value": "key1",
    },
    EventAttribute {
      "index": false,
      "key": "key2",
      "value": "key2",
    },
    EventAttribute {
      "index": false,
      "key": "key3",
      "value": "key3",
    },
  ],
  "type": "type",
}
`;

exports[`Event Recorder > should capture events 2`] = `
Event {
  "attributes": [
    EventAttribute {
      "index": false,
      "key": "key1",
      "value": "key1",
    },
    EventAttribute {
      "index": false,
      "key": "key2",
      "value": "key2",
    },
    EventAttribute {
      "index": false,
      "key": "key3",
      "value": "key3",
    },
  ],
  "type": "type",
}
`;
